portfolio/src/main/java/com/google/sps/data/CuisineDatastore.java

```diff
@@ -0,0 +1,57 @@
+package com.google.sps.data;
+
+import java.util.Map;
+import com.google.appengine.api.datastore.*;
+import com.google.common.collect.ImmutableMap;
+import com.google.sps.data.Constants;
+
+/** Class that handles writing/reading cuisine data from the datastore. */
+public final class CuisineDatastore {
+    
+    /** Method that retreives cuisine entities from the datastore and formats it. */
+    public static Map<String, Integer> fetchCuisineVotes() {
+      Query query = new Query(Constants.CUISINE_ENTITY);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      PreparedQuery results = datastore.prepare(query);
+      
+      ImmutableMap.Builder<String, Integer> cuisineVotes = ImmutableMap.builder();
+      for (Entity entity : results.asIterable()) {
+        String cuisine = (String) entity.getProperty(Constants.CUISINE_PARAMETER);
+        Long votes = (Long) entity.getProperty(Constants.VOTES_PARAMETER);
+        cuisineVotes.put(cuisine, votes.intValue());
+      }
+
+      return cuisineVotes.build();
+    }
+
+    /** Method that updates the cuisine vote in the datastore.
+    Method throws {@link IllegalArgumentException} if the cuisine is empty or null */
+    public static void addCuisineVote(String cuisine) {
+      if (cuisine.isEmpty()) {
+        throw new IllegalArgumentException(Constants.CUISINE_EMPTY_ERROR);
+      }
+
+      Query.Filter filter = new Query.FilterPredicate(Constants.CUISINE_PARAMETER,
```

Comment:
do you have a unique constraint on your cuisine column? if your addCuisineVote is called in parallel it's guaranteed you will insert more than one. (which you already know about). In any case you need limit 1 or a unique constraint (preferred if possible).